src/plugins/api-proxy/util/receive.js

```diff
@@ -0,0 +1,250 @@
+ * @param {Array} the callback args for npm request - [response, body], where
+ * `response` is an `Http.IncomingMessage` and `body` is the body text of the
+ * response.
+ * @return responseObj the JSON-parsed text, possibly with error info
+ */
+export const makeParseApiResponse = query => ([response, body]) => {
+	const { value, error } = parseApiValue([response, body]);
+	const meta = {
+		...parseMetaHeaders(response.headers),
+		endpoint: query.endpoint,
+		statusCode: response.statusCode,
+	};
+
+	return {
+		value,
+		error,
+		meta,
+	};
+};
+/**
+ * Format apiResponse to match expected state structure
+ *
+ * @param {Object} apiResponse JSON-parsed api response data
+ */
+export const makeApiResponseToQueryResponse = query => ({
+	value,
+	error,
+	meta,
+}) => ({
+	type: query.type,
+	ref: query.ref,
+	value,
+	error,
+	meta,
+});
+
+export const makeLogResponse = request => ([response, body]) => {
+	const {
+		elapsedTime,
+		request: { id, uri: { query, pathname, href }, method },
+		statusCode,
+	} = response;
+
+	const logger = request.server.app.logger;
+	// production logs will automatically be JSON-parsed in Stackdriver
+	const log = ((statusCode >= 400 && logger.error) ||
+		(statusCode >= 300 && logger.warn) ||
+		logger.info)
+		.bind(logger);
+
+	log(
+		{
+			type: 'response',
+			direction: 'in',
+			info: {
+				url: href,
+				query: (query || '').split('&').reduce((acc, keyval) => {
+					const [key, val] = keyval.split('=');
+					acc[key] = val;
+					return acc;
+				}, {}),
+				method,
+				id,
+				originRequestId: request.id,
+				statusCode: statusCode,
+				time: elapsedTime,
+				body: body.length > 256 ? `${body.substr(0, 256)}...` : body,
+			},
+		},
+		`Incoming response ${method.toUpperCase()} ${pathname} ${response.statusCode}`
+	);
+};
+
+/**
+ * When a tough-cookie cookie jar is provided, forward the cookies along with
+ * the overall /api response back to the client
+ */
+export const makeInjectResponseCookies = request => ([response, _, jar]) => {
+	if (!jar) {
+		return;
+	}
+	const requestUrl = response.toJSON().request.uri.href;
+	jar.getCookies(requestUrl).forEach(cookie => {
+		const cookieOptions = {
+			domain: cookie.domain,
+			path: cookie.path,
+			isHttpOnly: cookie.httpOnly,
+			isSameSite: false,
+			isSecure: request.server.settings.app.isProd,
+			strictHeader: false, // Can't enforce RFC 6265 cookie validation on external services
+		};
+
+		request.plugins[API_PROXY_PLUGIN_NAME].setState(
+			cookie.key,
+			cookie.value,
+			cookieOptions
+		);
+	});
+};
+
+export const makeReceive = request => {
+	const logResponse = makeLogResponse(request);
+	const injectResponseCookies = makeInjectResponseCookies(request);
+	return query => {
+		const parseApiResponse = makeParseApiResponse(query);
+		const apiResponseToQueryResponse = makeApiResponseToQueryResponse(query);
+		return response => {
+			logResponse(response); // this will leak private info in API response
```

Comment:
It's still pretty useful because it's the simplest way to what data is coming from the REST API for a particular request - Ivy used it just today to debug an issue, and I think it's probably used pretty regularly in dev. The privacy concern is partly limited by the fact that we truncate the 'body' of the response to 256 characters, and I think it's probably okay but imperfect to have this data in the logs - in theory it's as protected as any of our other prod data (and less of a privacy concern than QLing in my opinion), and you have to do some specific log parsing to expose it more generally. Definitely happy to have a followup discussion or just bring it up in Slack or something